scalyr_agent/copying_manager/worker.py

```diff
@@ -1021,29 +1033,53 @@ def __scan_for_new_bytes(self, current_time=None):
         for processor in self.__log_processors:
             processor.scan_for_new_bytes(current_time)
 
-    def __write_checkpoint_state(
-        self, log_processors, base_file, current_time, full_checkpoint
-    ):
-        # type: (List[LogFileProcessor], six.text_type, float, bool) -> None
-        """Writes the current checkpoint state to disk.
+    def _remove_closed_log_processors_and_save_checkpoints(self):
+        new_log_processors = []
 
-        This must be done periodically to ensure that if the agent process stops and starts up again, we pick up
-        from where we left off copying each file.
+        for processor in self.__log_processors:
+            if processor.is_closed():
+                self._closed_log_processors.append(processor)
```

Comment:
This should be protected by the lock, correct?  I would build a local variable with all of the closed log processors you need to add, and then add them at the end.

Reply:
This variable was removed https://github.com/scalyr/scalyr-agent-2/pull/734#discussion_r607252502